File: src/main/java/dev/stashy/extrasounds/debug/DebugUtils.java
package dev.stashy.extrasounds.debug;

import dev.stashy.extrasounds.mapping.SoundPackLoader;
import net.minecraft.entity.effect.StatusEffect;
import net.minecraft.entity.effect.StatusEffectCategory;
import net.minecraft.sound.SoundEvent;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.stream.Collectors;

public class DebugUtils
{
    public static final String debugVar = "extrasounds.debug";
    public static final String debugPathVar = "extrasounds.debug.path";

    public static final boolean debug = System.getProperties().containsKey(debugVar)
            && System.getProperty(debugVar).equals("true");
    public static final String debugPath = System.getProperties().containsKey(debugPathVar)
            ? System.getProperty(debugPathVar) : "debug/";

    private static final Logger LOGGER = LogManager.getLogger();

    public static void init()
    {
        if (!debug) return;
        LOGGER.info("ExtraSounds: DEBUG mode enabled.");
        LOGGER.info("Debug path: " + Path.of(debugPath).toAbsolutePath());
    }

    public static void exportSoundsJson(byte[] jsonData)
    {
        if (!debug) return;
        try
        {
            Path p = Path.of(debugPath).resolve("sounds.json");
            createFile(p);
            Files.write(p, jsonData, StandardOpenOption.TRUNCATE_EXISTING);
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
    }

    public static void exportGenerators()
    {
        if (!debug) return;
        Path p = Path.of(debugPath).resolve("generators.txt");
        createFile(p);
        try
        {
            Files.write(p, SoundPackLoader.mappers.keySet().stream()
                                                  .map(it -> {
                                                      var clazz = SoundPackLoader.mappers.get(it).itemSoundGenerator()
                                                                                         .getClass();
                                                      return "namespace: " + it + "; class: " + (clazz == null ? "none" : clazz.getName());
                                                  })
                                                  .collect(
                                                          Collectors.toList()));
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
    }

    public static void soundLog(SoundEvent snd)
    {
        if (!debug) return;
        if (snd.getId().getPath().startsWith("item.click"))
            LOGGER.info("Playing sound: " + snd.getId());
    }

    public static void effectLog(StatusEffect effect, boolean add)
    {
        if (!debug) return;
        boolean positive = !effect.getCategory().equals(StatusEffectCategory.HARMFUL);
        LOGGER.info(
                (positive ? "Positive" : "Negative") + " effect " + (add ? "added" : "removed") + ": " + effect.getName()
                                                                                                               .getString());
    }

    private static void createFile(Path p)
    {
        try
        {
            if (!Files.isDirectory(Path.of(debugPath)))
                Files.createDirectory(Path.of(debugPath));
            if (!Files.exists(p))
                Files.createFile(p);
        }
        catch (IOException e)
        {
            LOGGER.error("Unable to create file: " + p);
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/dev/stashy/extrasounds/DebugUtils.kt
package dev.stashy.extrasounds

import dev.stashy.extrasounds.mapping.SoundLoader
import net.minecraft.entity.effect.StatusEffect
import net.minecraft.entity.effect.StatusEffectCategory
import net.minecraft.sound.SoundEvent
import org.apache.logging.log4j.LogManager
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption


object DebugUtils {
    const val debugVar = "extrasounds.debug"
    const val debugPathVar = "extrasounds.debug.path"
    const val noCacheVar = "extrasounds.nocache"

    val debug = System.getProperties()[debugVar]?.equals("true") ?: false
    val debugPath = System.getProperties()[debugPathVar]?.toString() ?: "debug/"
    val skipCache = System.getProperties()[noCacheVar]?.equals("true") ?: false

    val logger = LogManager.getLogger()

    fun init() {
        if (!debug) return
        logger.info("ExtraSounds DEBUG mode enabled.")
        logger.info("Debug path: ${Path.of(debugPath).toAbsolutePath()}")
    }

    fun genericLog(msg: String) {
        if (!debug) return
        logger.info(msg)
    }

    fun soundLog(snd: SoundEvent) {
        if (!debug) return
        logger.info("Playing sound: ${snd.id}")
    }

    fun effectLog(effect: StatusEffect, add: Boolean) {
        if (!debug) return
        val positive = !effect.category.equals(StatusEffectCategory.HARMFUL)
        logger.info("${if (positive) "Positive" else "Negative"} effect ${if (add) "added" else "removed"}: ${effect.name.string}")
    }

    fun exportSoundsJson(data: ByteArray) {
        if (!debug) return
        try {
            val p = Path.of(debugPath).resolve("sounds.json")
            createFile(p)
            Files.write(p, data, StandardOpenOption.TRUNCATE_EXISTING)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun exportGenerators() {
        if (!debug) return
        val p = Path.of(debugPath).resolve("generators.txt")
        createFile(p)
        try {
            Files.write(p, SoundLoader.mappers.keys.stream()
                .map { it ->
                    val clazz = SoundLoader.mappers[it]?.javaClass
                    "namespace: " + it + "; class: " + if (clazz == null) "none" else clazz.name
                }
                .toList())
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun createFile(p: Path) {
        try {
            if (!Files.isDirectory(Path.of(debugPath))) Files.createDirectory(Path.of(debugPath))
            if (!Files.exists(p)) Files.createFile(p)
        } catch (e: IOException) {
            logger.error("Unable to create file: $p")
            e.printStackTrace()
        }
    }
}